marker_api/src/ast/common/span.rs

```diff
@@ -2,133 +2,469 @@ use std::marker::PhantomData;
-    pub fn snippet(&self) -> Option<String> {
-        with_cx(self, |cx| cx.span_snipped(self))
+    /// Returns a new [`Span`] with the given end position.
+    #[must_use]
+    pub fn with_end(&self, end: SpanPos) -> Span<'ast> {
+        let mut new_span = self.clone();
+        new_span.set_end(end);
+        new_span
     }
 
-    /// Converts a span to a code snippet if available, otherwise returns the default.
-    ///
-    /// This is useful if you want to provide suggestions for your lint or more generally, if you
-    /// want to convert a given [`Span`] to a [`String`]. To create suggestions consider using
-    /// [`snippet_with_applicability()`](`Self::snippet_with_applicability`) to ensure that the
-    /// [`Applicability`] stays correct.
-    ///
-    /// # Example
-    /// ```rust,ignore
-    /// // Given two spans one for `value` and one for the `init` expression.
-    /// let value = Vec::new();
-    /// //  ^^^^^   ^^^^^^^^^^
-    /// //  span1   span2
-    ///
-    /// // The snipped call would return the corresponding code snippets
-    /// span1.snippet_or("..") // -> "value"
-    /// span2.snippet_or("..") // -> "Vec::new()"
-    /// ```
-    pub fn snippet_or(&self, default: &str) -> String {
-        self.snippet().unwrap_or_else(|| default.to_string())
+    #[must_use]
+    pub fn source(&self) -> SpanSource<'ast> {
+        with_cx(self, |cx| cx.span_source(self))
     }
+}
 
-    /// Same as [`snippet()`](`Self::snippet`), but adapts the applicability level by following
-    /// rules:
-    ///
-    /// - Applicability level [`Unspecified`](`Applicability::Unspecified`) will never be changed.
-    /// - If the span is inside a macro, change the applicability level to
-    ///   [`MaybeIncorrect`](`Applicability::MaybeIncorrect`).
-    /// - If the default value is used and the applicability level is
-    ///   [`MachineApplicable`](`Applicability::MachineApplicable`), change it to
-    ///   [`HasPlaceholders`](`Applicability::HasPlaceholders`)
-    pub fn snippet_with_applicability(&self, default: &str, applicability: &mut Applicability) -> String {
-        if *applicability != Applicability::Unspecified && self.is_from_macro() {
-            *applicability = Applicability::MaybeIncorrect;
+#[cfg(feature = "driver-api")]
+impl<'ast> Span<'ast> {
+    #[must_use]
+    pub fn new(source_id: SpanSrcId, from_expansion: bool, start: SpanPos, end: SpanPos) -> Self {
+        Self {
+            _lifetime: PhantomData,
+            source_id,
+            from_expansion,
+            start,
+            end,
         }
-        self.snippet().unwrap_or_else(|| {
-            if *applicability == Applicability::MachineApplicable {
-                *applicability = Applicability::HasPlaceholders;
-            }
-            default.to_string()
-        })
+    }
+
+    pub fn source_id(&self) -> SpanSrcId {
+        self.source_id
+    }
+}
+
+#[repr(C)]
+#[derive(Debug)]
+#[allow(clippy::exhaustive_enums)]
```

Comment:
This enum should actually be `[non_exhaustive]` I just pasted the wrong thing :sweat_smile: 

For `[non_exhaustive]`, Marker's goal is to create a stable and future-proof API. It can always happen that we need to add new variants to enums. That's also the reason I enabled the restriction lint. With application code, it's something different, since you also control all consumers of the code you write. Marker is just a library which might also be used for closed source lints. So API changes can not be taken lightly, after v1.0.0